repair/repair.cc

```diff
@@ -1081,6 +1076,80 @@ future<int> repair_service::do_repair_start(sstring keyspace, std::unordered_map
+        options.column_families.size() ? options.column_families : list_column_families(db, keyspace);
+    if (cfs.empty()) {
+        rlogger.info("repair[{}]: completed successfully: no tables to repair", id.uuid());
+        co_return id.id;
+    }
+
+    {
+        std::sort(cfs.begin(), cfs.end());
+        size_t nr_tablet_table = 0;
+        size_t nr_vnode_table = 0;
+        bool is_tablet = false;
+        for (auto& table_name : cfs) {
+            auto& t = db.find_column_family(keyspace, table_name);
+            if (t.uses_tablets()) {
+                nr_tablet_table++;
+            } else {
+                nr_vnode_table++;
+            }
+        }
+
+        if (nr_tablet_table !=0) {
+            if (nr_vnode_table !=0) {
+                throw std::runtime_error("Mixed vnode table and tablet table");
+            }
+            is_tablet = true;
+        }
+        if (is_tablet) {
+            // Reject unsupported options for tablet repair
+            if (!options.ranges.empty()) {
+                throw std::runtime_error("The ranges option is not supported for tablet repair");
+            }
+            if (!options.hosts.empty()) {
+                throw std::runtime_error("The hosts option is not supported for tablet repair");
+            }
+            if (!options.ignore_nodes.empty()) {
+                throw std::runtime_error("The ignore_nodes option is not supported for tablet repair");
+            }
+            if (!options.data_centers.empty()) {
+                throw std::runtime_error("The dataCenters option is not supported for tablet repair");
+            }
+            if (!options.start_token.empty()) {
+                throw std::runtime_error("The starToken option is not supported for tablet repair");
+            }
+            if (!options.end_token.empty()) {
+                throw std::runtime_error("The endToken option is not supported for tablet repair");
+            }
+            if (options.small_table_optimization) {
+                throw std::runtime_error("The small_table_optimization option is not supported for tablet repair");
+            }
+
+            auto host2ip = [&ss = _ss] (locator::host_id host) -> future<gms::inet_address> {
+                return ss.invoke_on(0, [host] (service::storage_service& s) {
+                    return s.host2ip(host);
+                });
+            };
+            bool primary_replica_only = options.primary_range;
+            auto start_time = std::chrono::steady_clock::now();
+            for (auto& table_name : cfs) {
+                co_await repair_tablets(id, keyspace, table_name, host2ip, primary_replica_only);
+            }
+            auto duration = std::chrono::duration<float>(std::chrono::steady_clock::now() - start_time);
+            rlogger.info("repair[{}]: Finished user-requested repair for tablet keyspace={} tables={} repair_id={} duration={}", id.uuid(), keyspace, cfs, id.id, duration);
+            co_return id.id;
```

Comment:
This changes the semantic of the API, which now waits for the repair to finish instead of just starting the task.

It could interact badly with scylla manager. \cc @karol-kokoszka

Reply:
We need to fix the PR to make it following the old semantics.

Reply:
Fixed in v2.